### PR TITLE
Add support for ReasonML

### PR DESCRIPTION
**Pull request by TristanCacqueray**

### CODE REVIEW

**Comment by lassik** [comment]
Great, thanks!

Should the formatter be named `refmt` or `bsrefmt`? There's a forum discussion about it [here](https://github.com/lassik/emacs-language-id). It seems those two commands should both run the same formatter, but may run a different version of it, giving slightly different formatting results in some cases.

File: format-all.el [hunk 1]
```diff
@@ -623,6 +624,12 @@ Consult the existing formatters for examples of BODY."
   (:languages "PureScript")
   (:format (format-all--buffer-easy executable "-")))
 
+(define-format-all-formatter refmt
+  (:executable "bsrefmt")
+  (:install "npm install --global bs-platform")
+  (:languages "ReasonML")
```
Comment (lassik) [inline format-all.el:630]: Please change this from `"ReasonML"` to `"Reason"`. `"Reason"` is recognized by the GitHub Linguist project, so I added it to the Emacs [language-id](https://github.com/lassik/emacs-language-id) package which format-all uses.

File: format-all.el [hunk 2]
```diff
@@ -737,6 +744,7 @@ unofficial languages IDs are prefixed with \"_\"."
       (and (equal major-mode 'gleam-mode) "_Gleam")
       (and (equal major-mode 'ledger-mode) "_Ledger")
       (and (equal major-mode 'snakemake-mode) "_Snakemake")
+      (and (equal major-mode 'reason-mode) "ReasonML")
```
Comment (lassik) [inline format-all.el:747]: This line is no longer needed now that the language-id package covers Reason.

**Comment by lassik** [comment]
Perhaps we should call it `bsrefmt` in format-all. That is less ambiguous. We can add another formatter `refmt` using that specific command. If `bsrefmt` and `refmt` are later merged upstream, we can remove `bsrefmt` from format-all and keep only `refmt`.

File: format-all.el (hunk 1 above)
Comment (TristanCacqueray) [reply format-all.el:630]: That's good to know :-)

File: format-all.el (hunk 2 above)
Comment (TristanCacqueray) [reply format-all.el:747]: Done

**Comment by lassik** [comment]
Thanks - I tested that this works; merging.

Yet another project has popped up in this OCaml-to-JavaScript niche - ReScript: https://rescript-lang.org/blog/a-note-on-bucklescripts-future-commitments Does this leave the future of `bs-platform` and hence `bsrefmt` uncertain?

In any case, we can have `bsrefmt` in format-all now to solve the immediate problem. If they change their command name, we can change the name of the format-all formatter to match.

**Comment by lassik** [comment]
Is this the right homepage for bsrefmt: https://github.com/glennsl/bs-refmt

**Comment by TristanCacqueray** [comment]
If I understood correctly, ReScript is the merge of Reason (the language) and Bucklescript (the compiler), but Reason still exists as a standalone language that can be used for "native" project.

The ReScript team said they are committed to support the existing Reason syntax and `bs-platform` is still the recommended way to get the Reason formatter `bsrefmt`. The formatter is now provided by https://github.com/rescript-lang/rescript-compiler , it's actually also a transpiller that can convert Reason to OCaml too.

It seems like `bs-refmt` is another implementation to be used from Reason directly.

Needless to say this is all quite confusing but I don't think this should affect `format-all` in the near future. Thank you for testing and accepting the change!

**Comment by lassik** [comment]
Thanks for explaining it. I'm still confused as to whether or not the BuckleScript brand name and the `bs-platform` npm package will stick around. Is `bs-refmt` (with a dash) yet another variation on the formatter as distinguished from `refmt` and `bsrefmt` (no dash)?

In any case, the best thing to do in format-all is probably to follow their naming precisely, adding/removing/changing names as they do. Any other approach is likely to be even more confusing :)

**Comment by TristanCacqueray** [comment]
If I understood correctly (and please note I am pretty new to the ecosystem), the BuckleScript brand is being deprecated and the npm package will be renamed to `rescript` see this [forum post](https://forum.rescript-lang.org/t/will-there-be-a-new-npm-package/53). I can't tell how the `bs-refmt` is implemented and it seems like `refmt` came from https://github.com/reasonml/reason-cli which has been deprecated by `bs-platform`.

Also note that `bsrefmt` is actually a wrapper for an embedded `refmt.exe` binary from the `bs-platform` package. Looking at the vscode extension, it directly refers to that binary from the `bs-platform` [here](https://github.com/jaredly/reason-language-server/blob/b28eb59f47d17481c431165a0cb3e0c5bc773166/src/analyze/BuildSystem.re#L357).

I think it's better to name it `bsrefmt` since its what is installed and this also avoid confusion with the legacy `refmt`.

**Comment by lassik** [comment]
Wow, what a maze of projects, packages, names and repos :) Let's hope things eventually settle down.

> I think it's better to name it `bsrefmt` since its what is installed and this also avoid confusion with the legacy `refmt`.

Sounds good. The install command we currently recommend is `npm install --global bs-platform`. We can change it once `bs-platform` is deprecated and a new alternative is solid.

**Pull request merged**